Use number type for user related operators.

amanagr opened the pull request.

amanagr committed:
compose_closed_ui: Avoid error for invalid user ids.

It is possibly for user to incorrectly paste a user id or type one,
we don't need to throw errors in that case and just show our
default label there.

evykassirer commented on web/src/filter.ts:
```diff
@@ -172,16 +172,18 @@ function message_matches_search_term(message: Message, term: NarrowTerm): boolea
-            if (!operand_ids) {
-                return false;
+
+            let operand_ids: number[] = term.operand;
```
Why not use `sorted_other_user_ids` here?

amanagr replied:
Forgot to use that here. Done.

evykassirer commented on web/src/filter.ts:
```diff
@@ -354,10 +350,14 @@ export class Filter {
-            return operand.replaceAll(/[\s"%]/g, (c) => encodeURIComponent(c));
+    static encodeOperand(operand: NarrowTerm["operand"]): string {
+        if (typeof operand !== "string") {
+            return String(operand);
```
I guess the other types of `operand` don't need to be encoded?

evykassirer commented on web/src/filter.ts:
```diff
@@ -460,30 +460,69 @@ export class Filter {
+                case "sender":
+                    potential_narrow_term = {
+                        operator: canonical_operator,
+                        operand:
```
Especially here, because then you could return early after calculating it, before building `potential_narrow_term`

evykassirer commented on web/src/filter.ts:
```diff
@@ -515,11 +554,10 @@ export class Filter {
-                return term.operand
-                    .split(",")
-                    .every((email) => people.get_by_email(email) !== undefined);
+            case "dm":
```
This is small, but moving the `case "dm"` line back up would clean up the diff

amanagr replied:
Done and used `people.is_valid_user_ids(term.operand);`.

evykassirer commented on web/src/filter.ts:
```diff
@@ -1271,11 +1308,7 @@ export class Filter {
-                                .map((term) => term.operand)
-                                .join(","),
-                        )
+                        people.user_ids_to_slug(this.terms_with_operator("dm")[0]!.operand)
```
Oh hm, there's a lot of `[0]!` below. That would be nice to clean up with `util.the` too, probably, though that can be a followup.

evykassirer commented on web/src/hash_util.ts:
```diff
@@ -112,7 +126,7 @@ export function by_stream_topic_url(stream_id: number, topic: string): string {
 // corresponding hash: the # component
 // of the narrow URL
-export function search_terms_to_hash(terms?: NarrowTerm[]): string {
+export function search_terms_to_hash(terms?: NarrowCanonicalTerm[]): string {
```
Can you explain more why the changes to this function happened as a result of moving to numbers for user operators?

evykassirer commented on web/src/hash_util.ts:
```diff
@@ -375,14 +385,12 @@ export function decode_dm_recipient_user_ids_from_narrow_url(narrow_url: string)
-            if (!user_ids) {
-                return null;
-            }
+        assert(terms[0].operator === "dm");
```
Given the above `if`, it seems like `terms[0].operator !== "dm"` sometimes. Why do you know that's not the case at this point?

evykassirer replied:
I would say then to add the `assert(terms[0].operator === "dm");` in a prep commit, and also to remove the above `if` block that has `terms[0].operator !== "dm" &&` in it.

Also since this is all in a `try` block, if the `assert` fails, the function will quietly return `null`. I'd say either return `null` here to do that more explicitly, or make the assert outside of the `try`

evykassirer commented on web/src/hash_util.ts:
```diff
@@ -375,14 +385,12 @@ export function decode_dm_recipient_user_ids_from_narrow_url(narrow_url: string)
             return user_ids;
         }
-        return null;
+        return terms[0].operand;
```
Why this change from `null` to `terms[0].operand`?

amanagr replied:
Oops! Changed.

evykassirer commented on web/src/people.ts:
```diff
@@ -789,30 +757,8 @@ export function user_ids_to_full_names_array(user_ids: number[]): string[] {
-
 export function user_ids_to_slug(user_ids: number[]): string | undefined {
-    if (user_ids.length === 0) {
+    if (user_ids.length === 0 || !user_ids.every((id) => is_valid_user_id(id))) {
```
`is_valid_user_ids` seems like it would be helpful here

evykassirer commented on web/src/pm_list_data.ts:
```diff
@@ -20,17 +20,13 @@ export function get_active_user_ids_string(): string | undefined {
+    const user_ids = filter.terms_with_operator("dm")[0]?.operand;
 
-    if (!emails) {
+    if (user_ids?.length === 0 || !user_ids?.every((id) => people.is_valid_user_id(id))) {
```
I think `if (!user_ids || user_ids.length === 0 || !people.is_valid_user_ids(user_ids)) {` would be nicer here, but that might just be personal preference

amanagr replied:
Yeah, changed to use your version.

evykassirer commented on web/src/util.ts:
```diff
@@ -271,16 +271,6 @@ export function is_channels_synonym(text: string): boolean {
     return text === "streams";
 }
 
-export function canonicalize_channel_synonyms(text: string): string {
```
Why is this a consequence of moving user operators to number type? Or could it be a prep commit?

evykassirer commented on web/src/search_pill.ts:
```diff
@@ -55,9 +54,11 @@ export function create_item_from_search_string(search_string: string): SearchPil
-    return `${sign}${item.operator}: ${get_search_operand(item, true)}`;
+export function get_search_string_from_narrow_suggestion(suggestion: NarrowTermSuggestion): string {
+    // Call `get_search_string_for_copying` to get `operand` before calling this.
+    // We don't do that here to avoid doing extra work to please typescript.
```
Thoughts on doing this instead? It feels simpler. I'm having trouble following why you're changing other things when calling this.

```js
export function get_search_string_from_narrow_suggestion(item: SearchPill): string {
    function get_search_string_for_copying(item: SearchPill): string {
        switch (item.operator) {
            case "dm":
            case "dm-including":
            case "sender":
                assert(item.type === "search_user");
                return item.users.map((user) => user.email).join(",");
            case "topic":
                return util.get_final_topic_display_name(item.operand);
            case "channel":
                if (item.operand === "") {
                    return "";
                }
                return stream_data.get_valid_sub_by_id_string(item.operand).name;

            default:
                return item.operand;
        }
    }

    const suggestion = {
        operator: item.operator,
        operand: get_search_string_for_copying(item),
        negated: item.negated,
    }
    
    const sign = suggestion.negated ? "-" : "";
    return `${sign}${suggestion.operator}: ${suggestion.operand}`;
}
```

evykassirer commented on web/src/search_pill.ts:
```diff
@@ -247,26 +261,36 @@ export function create_pills($pill_container: JQuery): SearchPillWidget {
+                    return render_input_pill({
+                        display_value: get_pill_display_value(item),
+                    });
+                default:
```
Maybe `assert(item.type !== "search_user");` here just in case?

evykassirer commented on web/src/search_pill.ts:
```diff
@@ -275,14 +299,22 @@ export function create_pills($pill_container: JQuery): SearchPillWidget {
-        assert(person !== undefined);
+    const user_ids = get_user_ids_from_term_with_user_pill_operator(term);
+    const users = user_ids.map((user_id) => {
+        const person = people.get_by_user_id(user_id);
```
`const users = user_ids.map((user_id) => people.get_by_user_id(user_id));`

evykassirer commented on web/src/search_suggestion.ts:
```diff
@@ -349,14 +355,15 @@ function get_group_suggestions(
         // Take top 15 persons, since they're ordered by pm_recipient_count.
         persons = persons.slice(0, 15);
 
+        const user_ids = people.emails_string_to_user_ids(all_but_last_part);
```
Why not change `all_but_last_part` to be user ids?

evykassirer commented on web/src/search_suggestion.ts:
```diff
@@ -942,8 +963,35 @@ function get_operator_suggestions(
-        const op = [{operator: choice, operand: "", negated}];
-        return format_as_suggestion(op);
+
+        // Operand is ignored but we need to respect typescript type checking.
```
Can you add a more clear comment here? I had to do some poking around to figure out what was going on.

Something like: 

> "Set `is_operator_suggestion` to `true`, since we're only suggesting the operator and don't want any operand, but we can't put empty strings here for typescript reasons.

Hm, alternatively, in `unparse`, you could check for these operators with empty operands and use an empty string instead. I'm not sure where it's least confusing to put the boundary between types.

amanagr replied:
Added the comment.

evykassirer reviewed (commented):
I've looked at the main files but not tests yet! Will look at those on the next pass hopefully. Thanks for all the work you've done on this, what a doozy! I left several questions and comments, feel free to ask if you have followup questions on any of them.

amanagr commented on web/src/filter.ts:
```diff
@@ -192,10 +194,7 @@ function message_matches_search_term(message: Message, term: NarrowTerm): boolea
-            const operand_ids = people.pm_with_operand_ids(term.operand);
-            if (!operand_ids) {
-                return false;
-            }
```
This is new since we are no longer doing the `email` to `user` to `user_id` converstion.

evykassirer replied:
Ah right, because `sorted_other_user_ids` can't return `undefined`

amanagr commented on web/src/filter.ts:
```diff
@@ -240,6 +239,9 @@ export class Filter {
     }: NarrowTerm): NarrowCanonicalTerm {
+        // NOTE: This function is for conversion of `NarrowTerm` to
+        // `NarrowCanonicalTerm`. For more general conversion, use
+        // `convert_suggestion_to_term`.
```
Changed to use this, thanks for noticing.

// NOTE: For conversion of NarrowTermSuggestion to NarrowCanonicalTerm,
// use `Filter.convert_suggestion_to_term` instead.

amanagr commented on web/src/filter.ts:
```diff
@@ -676,7 +712,12 @@ export class Filter {
+        is_operator_suggestion: boolean,
+    ): Part[] {
+        // search_pill.generate_pills_html already does all the conversion
+        // of terms that are necessary, we don't need to it again here.
```
I don't know if there is way to make sure of that but doing the conversion of terms here is not easy due to the custom text involved in operands.

Yes, that is the purpose of this comment.

evykassirer replied:
Ah, I see. I think having a comment here saying that more explicitly -- calling `canonicalize_term` on the terms is not easy here, and so it's expected that callers already do those conversions, and the current only callpath (generate_pills_html) does.

Though ideally there would be a way to get this typed as `terms: NarrowCanonicalTerm[]` to show that

amanagr commented on web/src/hash_util.ts:
```diff
@@ -65,13 +67,25 @@ export function encode_stream_id(stream_id: number): string {
+    if (operator === "sender") {
+        if (operand === "me") {
+            return people.my_current_user_id();
+        }
```
This was previously done in `canonicalize_term` but we have to do it here due to the return type of `parse_narrow` being `NarrowCanonicalTerm`.

evykassirer replied:
I'm not totally following, sorry. `parse_narrow` had that return type already yeah? So `canonicalize_term` was previously called somewhere before that it's not being called anymore?

amanagr replied:
`parse_narrow` already had `NarrowCanonicalTerm` return type but we change the definition of `NarrowCanonicalTerm` in this commit.

evykassirer replied:
Okay I see:

* `parse_narrow` calls `decode_operand` for the term's operand
* `parse_narrow` returns `z.array(narrow_canonical_term_schema).parse(terms);` both before and after this PR
* in this commit, `parse_narrow` can no longer return "me", so we need to decode it here

I think there should be a prep commit that handles this here and translates it to the email, and then change that to a number in this commit. I'm assuming it's also relevant for `dm` operator. 

If this allows you to remove the "me" checks elsewhere, that would be clarifying to remove them in that prep commit, too.

amanagr replied:
> I think there should be a prep commit that handles this here and translates it to the email, and then change that to a number in this commit. I'm assuming it's also relevant for dm operator.

Yes, nice catch! Done!

amanagr committed:
hash_util: Raise error for invalid sender operator in URL.

amanagr commented on web/src/narrow_state.ts:
```diff
@@ -204,17 +195,23 @@ export function pm_ids_string(current_filter: Filter | undefined = filter()): st
+        return undefined;
+    }
+
+    return String(pm_ids(current_filter));
```
Oops! Changed.

amanagr replied:
Some of these changes are so old that I ignored to re-check them. :(

amanagr commented on web/src/people.ts:
```diff
@@ -848,15 +794,14 @@ export function slug_to_emails(slug: string): string | undefined {
 }
 
-export function exclude_me_from_string(user_ids_string: string): string {
+export function exclude_me_from_string(user_ids_string: string): number[] {
```
>  you could have it take number[]

Yes, also renamed it to `exclude_me_from_user_ids`.

amanagr commented on web/src/search_pill.ts:
```diff
@@ -366,26 +398,27 @@ export function set_search_bar_contents(
             case "dm":
             case "dm-including":
-            case "sender":
-                if (term.operand !== "") {
```
This is due to `const narrow_term = Filter.convert_suggestion_to_term(term);`, so it cannot be a prep commit.

amanagr committed:
search_pill: Simplify type of `pill_render_data`.

`NarrowCanonicalTerm` is not required here since user operators
use `SearchUserPill` and hence, all other operands are string
by default. So, they can pass as `NarrowTermSuggestion`.

amanagr commented on web/src/search_pill.ts:
```diff
@@ -218,7 +226,13 @@ export function generate_pills_html(suggestion: Suggestion, text_query: string):
                 search_terms[0]!.operator !== "" && !text_query.includes(":");
             description_html = Filter.search_description_as_html(
-                search_terms,
+                [
```
Added a prep commit for this.

evykassirer replied:
Seems like this could also be a prep commit:

```diff
diff --git a/web/src/search_pill.ts b/web/src/search_pill.ts
index 2e676f08d8..e4c6d508ef 100644
--- a/web/src/search_pill.ts
+++ b/web/src/search_pill.ts
@@ -218,7 +218,13 @@ export function generate_pills_html(suggestion: Suggestion, text_query: string):
             const is_operator_suggestion =
                 search_terms[0]!.operator !== "" && !text_query.includes(":");
             description_html = Filter.search_description_as_html(
-                search_terms,
+                [
+                    {
+                        operator: render_data.operator,
+                        operand: render_data.operand,
+                        negated: render_data.negated,
+                    },
+                ],
                 is_operator_suggestion,
             );
             const capitalized_first_letter = description_html.charAt(0).toUpperCase();
```

amanagr replied:
Yeah, added a prep commit for this.

amanagr commented on web/src/search_pill.ts:
```diff
@@ -247,26 +261,36 @@ export function create_pills($pill_container: JQuery): SearchPillWidget {
-        get_text_from_item: get_search_string_from_item,
+        get_text_from_item(item) {
+            return get_pill_display_value(item);
+        },
```
Sorry, I didn't find any other discussion of `get_search_string_from_item` from you above.

amanagr replied:
Yeah, done.

amanagr commented on web/src/search_suggestion.ts:
```diff
@@ -284,14 +284,20 @@ function get_group_suggestions(
             const text_input = last.operand;
-            const operand = `${last_complete_term.operand},${text_input}`;
+            // We use emails to generate suggestions.
+            // Hence, it is important for `operand` to be represented as emails.
```
This is due to how this function is defined:

```
export function build_person_matcher(query: string): (user: User) => boolean {
    query = query.trim();

    const termlets = query.toLowerCase().split(/\s+/);
    const termlet_matchers = termlets.map((termlet) => build_termlet_matcher(termlet));

    return function (user: User): boolean {
        const email = user.email.toLowerCase();

        if (email.startsWith(query)) {
            return true;
        }

        return termlet_matchers.every((matcher) => matcher(user));
    };
}

```

Changing it requires a separate discussion, so can skip that part in this PR.

evykassirer replied:
Is there a CZO thread yet for that discussion? Seems worth starting

amanagr replied:
[#api design > defer loading user data @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/defer.20loading.20user.20data/near/1905551) has some context.

evykassirer replied:
I'll let @timabbott comment on that on his pass, then. Also relevant for code in `get_search_string_from_item` it seems.

amanagr replied:
I think I was confused about how this works. In `build_person_matcher` we are just looking for matches to `email` and not `delivery_email` which is what user would usually type in the search bar. 

So, we can switch to using user ids here.

evykassirer reviewed (commented):
I got partway through looking over it again! Will return to it later tonight or tomorrow.

evykassirer commented on web/src/search_pill.ts:
```diff
@@ -55,9 +54,35 @@ export function create_item_from_search_string(search_string: string): SearchPil
+    };
+
+    const sign = suggestion.negated ? "-" : "";
+    return `${sign}${suggestion.operator}: ${suggestion.operand}`;
```
Seems simpler to do this instead probably?

```js
    const sign = suggestion.negated ? "-" : "";
    const operand = get_search_string_for_copying(item);
    return `${sign}${item.operator}: ${operand}`;
```

evykassirer commented on web/src/search_pill.ts:
```diff
@@ -55,9 +54,35 @@ export function create_item_from_search_string(search_string: string): SearchPil
-export function get_search_string_from_item(item: SearchPill): string {
-    const sign = item.negated ? "-" : "";
-    return `${sign}${item.operator}: ${get_search_operand(item, true)}`;
+export function get_search_string_from_narrow_suggestion(item: SearchPill): string {
```
This should probably keep the name `get_search_string_from_item` (or you could call it `from_search_pill`, but that rename would be for a prep commit)

evykassirer commented on web/src/search_pill.ts:
```diff
@@ -408,31 +450,36 @@ export function set_search_bar_contents(
-        // For all other `for_display=true` cases, we just show the default operand.
-    }
-    return item.operand;
+function get_pill_display_value(item: SearchPill): string {
```
I think you can get rid of this function and just use `get_search_string_from_narrow_suggestion` directly

amanagr committed:
hash_util: Use canonical term type for search_terms_to_hash.

Verified that the callers provided the correct term type.

amanagr committed:
hash_util: Help typescript understand the type of first_term.

amanagr committed:
hash_util: Use correct type for `encode_operand`.

This help us remove an unnecessary function call.

Checked that the only caller provided us the correct parameter type.

amanagr committed:
search_pill: Use `render_data` to generate search description.

We want to use the parsed values in `render_data` instead of
`search_terms` to generate search description.

amanagr committed:
hash_util: Use canonical term to help typescript understand types.

This will be useful to create nice diffs when user operators will
use number type.

amanagr committed:
narrow_state: Return complete NarrowTerm instead of just operand.

This will help us retain context for the operand in the NarrowTerm,
which will help Typescript determine the type of operand when
used in the calling function.

amanagr commented on web/src/search_pill.ts:
```diff
@@ -247,26 +278,36 @@ export function create_pills($pill_container: JQuery): SearchPillWidget {
-            });
         },
-        get_display_value_from_item: get_search_string_from_item,
+        get_display_value_from_item: get_pill_display_value,
```
Yeah, done.

amanagr committed:
search_suggestion: Use canonical terms where possible.

We can directly mark them as cononical based on how they are built.

amanagr commented:
Ready for another review. @evykassirer can you please close the conversations that you think are resolved?

Also, thanks for the thorough reviews to get this simplified.

evykassirer commented on web/src/filter.ts:
```diff
@@ -460,30 +455,79 @@ export class Filter {
+        let potential_narrow_term: NarrowCanonicalTerm;
         try {
-            const potential_narrow_term: NarrowCanonicalTerm = Filter.canonicalize_term(suggestion);
+            const canonical_operator = filter_util.canonicalize_operator(suggestion.operator);
```
Ah, right. Okay that makes sense.

evykassirer commented on web/src/util.ts:
```diff
@@ -273,9 +273,12 @@ export function is_channels_synonym(text: string): boolean {
 
 export function canonicalize_channel_synonyms(text: string): string {
     if (is_channel_synonym(text.toLowerCase())) {
+        // Function is removed in later commits.
```
not ideal

evykassirer reviewed (commented):
Ok pretty much done the main files! Will look at tests tomorrow

evykassirer commented on web/tests/hashchange.test.cjs:
```diff
@@ -166,26 +121,18 @@ run_test("people_slugs", () => {
-    assert.deepEqual(narrow, [{operator: "dm", operand: "alice@example.com", negated: false}]);
-
-    // Even though we renamed "pm-with" to "dm", preexisting
-    // links/URLs with "pm-with" operator are handled correctly.
```
Are we not handling this anymore?

amanagr replied:
`Filter` library handles everything now, which already has tests for this. So, no need to handling them anywhere else.

evykassirer commented on web/tests/message_view.test.cjs:
```diff
@@ -866,44 +864,44 @@ run_test("narrow_to_compose_target direct messages", ({override, override_rewire
 
-    let emails;
-    override(compose_pm_pill, "get_emails", () => emails);
+    let get_user_ids;
```
I would keep this `user_ids`, since `get` makes it sound like it's a function not a return value.

evykassirer commented on web/tests/narrow_unread.test.cjs:
```diff
@@ -117,7 +122,7 @@ run_test("get_unread_ids", () => {
     assert_unread_info({flavor: "cannot_compute"});
 
-    terms = [{operator: "dm", operand: "123123"}];
+    terms = [{operator: "dm", operand: [123123]}];
```
Can you make a variable to make this more clear? Like `bogus_user_id = 123123`

evykassirer commented on web/tests/pm_list_data.test.cjs:
```diff
@@ -246,15 +247,13 @@ test("get_active_user_ids_string", () => {
+    set_pm_with_filter([bob.user_id, alice.user_id]);
     assert.equal(pm_list_data.get_active_user_ids_string(), "101,102");
 
-    blueslip.expect("warn", "Unknown emails");
```
Why no warn for unknown id?

amanagr replied:
I missed to convert the function correctly on this one. Fixed to raise an error.

evykassirer commented on web/tests/search.test.cjs:
```diff
@@ -174,21 +174,21 @@ run_test("initialize", ({override, override_rewire, mock_template}) => {
                         {
                             description_html: "group direct messages including",
-                            search_string: "dm-including:user7@zulipdev.com",
+                            search_string: "dm-including:3",
```
can you do

```js
const zoe = {
    email: "user7@zulipdev.com",
    user_id: 3,
    full_name: "Zoe",
}
people.add_active_user(zoe, "server_events");
```

And then use `zoe.user_id` here instead of 3?

amanagr replied:
Done.

amanagr committed:
hash_util: Handle `me` operand.

We want `parse_narrow` to use user_ids for user operators, so
as a prep commit, we are converting `me` to use `email` instead.

amanagr committed:
filter: Mark empty user terms as invalid.

amanagr commented on web/tests/message_view.test.cjs:
```diff
@@ -384,22 +383,21 @@ run_test("show_empty_narrow_message", ({mock_template, override}) => {
 
-    people.add_active_user(alice);
-    current_filter = set_filter([["dm", "alice@example.com,Yo"]]);
+    current_filter = set_filter([["dm", [9999, 8888]]]);
```
Doesn't matter here since we are just checking for atleast one invalid user here.

amanagr commented on web/tests/narrow_state.test.cjs:
```diff
@@ -344,14 +323,6 @@ test("pm_ids_string", () => {
-
-    set_filter([["dm", "bogus@foo.com"]]);
-    assert.equal(narrow_state.pm_ids_string(), undefined);
-    assert.deepStrictEqual(narrow_state.pm_ids_set(), new Set());
```
Oops, restored !

amanagr commented on web/tests/people.test.cjs:
```diff
@@ -966,22 +968,13 @@ run_test("multi_user_methods", () => {
     let emails_string = people.user_ids_string_to_emails_string("402,401");
     assert.equal(emails_string, "emp401@example.com,emp402@example.com");
 
-    emails_string = people.slug_to_emails("402,401");
```
Sure, done!

amanagr committed:
search_suggestion: Explain use of the condition.

amanagr committed:
people: Use visible email for matching user query.

User is unlikely to use our system generated email for searching
a user. So, we use user's original email for matching where possible.

amanagr committed:
people: Allow user to lookup users using user id.

amanagr committed:
filter: Use user_id instead email for user type operators.

For `dm`, `dm-including` and `sender` operators, we use user ids
as operands instead of emails.

evykassirer commented on web/src/hash_util.ts:
```diff
@@ -62,6 +62,10 @@ export function decode_operand(
 ): string {
+    if (operand === "me" && (operator === "dm" || operator === "sender")) {
+        return people.my_current_email();
+    }
```
Can you remove it from elsewhere when you add it here?

amanagr replied:
We cannot do that since we want support `dm:me` in search too.

evykassirer commented on web/tests/message_view.test.cjs:
```diff
@@ -961,17 +959,14 @@ run_test("narrow_compute_title", () => {
     assert.equal(narrow_title.compute_narrow_title(filter), "joe");
 
-    filter = new Filter([{operator: "dm", operand: "joe@example.com,sally@doesnotexist.com"}]);
-    blueslip.expect("warn", "Unknown emails");
```
I think it would be ideal to have as few changes in functionality in this PR as possible, but @timabbott can also weigh in when he reviews, or you could double check in a CZO thread.

evykassirer replied:
Or you can make a prep commit removing the warn!

evykassirer commented on web/tests/filter.test.cjs:
```diff
@@ -2692,11 +2665,7 @@ test("navbar_helpers", ({override}) => {
-                "sally@doesnotexist.com",
-                steve.full_name,
-            ]),
+            title: "9999 and joe",
```
Awesome!

amanagr committed:
narrow_title: Remove error when generating narrow title.

We don't need to throw error when generating narrow title, the
invalid narrow logic is already handled in hashchange.

timabbott commented on web/tests/people.test.cjs:
```diff
@@ -1922,6 +1922,7 @@ run_test("fetch_users", async ({override}) => {
     assert.ok(people.is_valid_user_id(16));
     assert.ok(people.is_valid_user_id(42));
     assert.ok(people.is_valid_user_id(17));
+    assert.ok(people.is_valid_user_ids([15, 16, 42, 17]));
```
Should include some test coverage where it fails.

timabbott commented on web/src/hash_util.ts:
```diff
@@ -71,6 +71,9 @@ export function decode_operand(
         if (emails) {
+            if (operator === "sender") {
+                assert(!emails.includes(","), "Sender operand should not be a group");
+            }
```
I don't understand doing an `assert` here -- this situation is bad user input (e.g., changing the operator in the URL bar and hitting Enter); that isn't a bug in the code, right? 

Testing, this change introduces a bug; previously, we showed "This sender doesn't exist" as the error message, and now we just show the Inbox view with a JS assertion firing.

timabbott commented:
And thanks @evykassirer for reviewing as well, this was a quite big transition to read through!

andersk commented on web/src/search_pill.ts:
```diff
@@ -56,8 +55,29 @@ export function create_item_from_search_string(search_string: string): SearchPil
+            operand = util.get_final_topic_display_name(item.operand);
+            break;
+        case "channel":
+            if (item.operand === "") {
+                operand = "";
+            }
+            operand = stream_data.get_valid_sub_by_id_string(item.operand).name;
```
The `operand = ""` assignment is unused, which is probably not what you meant.

amanagr replied:
Thanks for catching! Opened #39638 to remove this.
